[sqlite] add a way to provide custom directory for the database

IgorKhramtsov opened the pull request:
# Why

Hey, thank you for the awesome ecosystem and packages! In the app I'm working on, I needed a way to share the SQLite database with a native widget, on iOS this could be done using shared containers, but that requires a way to provide a path for the SQLite database. 

This PR resolves the following feature request:
* https://expo.canny.io/feature-requests/p/support-for-setting-sqlite-database-path-to-ios-app-groups-directory

And partially resolves this feature request (a new function should be introduced in the `expo-file-system` package to get the path to the iCloud directory):
* https://expo.canny.io/feature-requests/p/sqlite-choose-location

# How

This PR does a few things in order to provide a way to store a database in an iOS shared container:
1. The new `directory` parameter is added, which will be combined with the `databaseName` to create a full path before supplying it to the native code. This way the backward compatibility is kept without introducing new methods;
2. The new `getSharedContainer` function is added in the `expo-file-system/next` package to get the path to a shared container;
3. Additional logic is added in the `getPathPermissions` function for paths in shared containers on iOS. Apparently, the permission check works differently in the shared container so when you try to check permissions of non-existent directory or parent directory - it will fail although there are enough permissions on the file system. The new logic will just check permissions in the root of the shared container;
4. New `ensureDatabasePathExists` (Sync and Async) functions are added in the SQLite native code to verify that the path exists and there are enough permissions before opening the database. Those functions basically are part of the constructor code moved out as during the tests I noticed that exceptions thrown in the constructor are not propagated correctly to the user (instead, it says that it was unable to find the native shared object associated with the given JS object; looks like it tries to map native Exception to the JS NativeDatabase or smth like that).


# Test Plan

New tests are added in the `test-suite` app for
1. `expo-file-system` to verify that it can create/delete files in the shared iOS container (and there are no permission issues);
2. `expo-sqlite` to verify that the database can be created opened and deleted in the custom directory and in the shared iOS container (I've added an app group in the bare-expo app for this test, but that might not work in the CI/CD if tests are executed from another app).

I, then, verified that those tests pass on iOS 17, 16, and Android 14.0 simulators using `bare-expo` app.
Also, I have tested this functionality in the app I'm working on by publishing forks of those two packages and using them in my project.


# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

IgorKhramtsov committed:
feat(expo-sqlite): support optional app group on ios

IgorKhramtsov committed:
refactor: put appGroup into iosOptions

IgorKhramtsov committed:
Merge branch 'main' into ik/sqlite-ios-app-group

IgorKhramtsov committed:
fix: swift compile errors

IgorKhramtsov committed:
chore: revert package.json changes

IgorKhramtsov committed:
chore: put back Utilities.urlFrom

IgorKhramtsov committed:
fix: swap iosOptions param to not break existing code
chore: remove erroneous doc comments

IgorKhramtsov committed:
chore: put back plugin build files

IgorKhramtsov committed:
[sqlite][ios] throw custom error when can't open shared container
[sqlite][android] adjust native code signatures
[core][ios] add custom handling for permission check of path in a shared
container
[file-system][ios] add getSharedContainerUriAsync
[sqlite][ios] add tests for app group use
[file-system][ios] add tests for getSharedContainerUriAsync use

IgorKhramtsov committed:
[sqlite][ios] add a way to provide custom directory
[sqlite][ios] remove appGroup and iosOptions paramsand

IgorKhramtsov committed:
[sqlite][ios] use databasePath in the native API, combine directory and
databaseName in the TS API

IgorKhramtsov committed:
[sqlite][android] add support for custom directory

IgorKhramtsov committed:
Merge branch 'main' into ik/sqlite-custom-directory

IgorKhramtsov committed:
[sqlite][ios] fix tests

IgorKhramtsov committed:
[sqlite] remove erroneous function declarations

IgorKhramtsov committed:
[sqlite] put back erroneously removed constructor param

IgorKhramtsov committed:
[sqlite] update CHANGELOG
[file-system] update CHANGELOG

IgorKhramtsov committed:
[core] update CHANGELOG

IgorKhramtsov committed:
fix swiftlint warnings

IgorKhramtsov committed:
[doc] example of storing sqlite db in a shared container

IgorKhramtsov commented on packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModuleNext.kt:
```diff
@@ -187,16 +194,20 @@ class SQLiteModuleNext : Module() {
+      ensureDirExists(parentPath)
+      return path.canonicalPath
+    } catch (e: IOException) {
+      throw OpenDatabaseException(databasePath, e.message)
```
Previously the original `IOException` message was lost but it could be helpful for debugging/reporting

IgorKhramtsov commented on packages/expo-sqlite/ios/SQLiteModuleNext.swift:
```diff
@@ -38,52 +38,54 @@ public final class SQLiteModuleNext: Module {
+      let path = try ensureDatabasePathExists(path: databasePath)
       let fileManager = FileManager.default
-      if fileManager.fileExists(atPath: path.absoluteString) && !forceOverwrite {
+      if fileManager.fileExists(atPath: path.standardizedFileURL.path) && !forceOverwrite {
```
Here and below `absoluteString` is replaced with `standardizedFileUrl.path` because otherwise, it fails if the path points to the shared container

IgorKhramtsov committed:
chore(sqlite): fix erroneous signatures

Kudo commented on packages/expo-sqlite/package.json:
```diff
@@ -51,7 +51,8 @@
   "dependencies": {
-    "@expo/websql": "^1.0.1"
+    "@expo/websql": "^1.0.1",
+    "expo-file-system": "^17.0.1"
```
having expo-file-system dependency may bring multiple expo-file-system versions in node_modules if people don't setup version correctly. i would recommend not introducing the other expo modules dependencies inside an expo module.

i looked into your pr and it turns out you were using expo-file-system only to get the document directory path. how about serving the feature from expo-sqlite directly? you can use [Constants](https://docs.expo.dev/modules/module-api/#constants)

Kudo reviewed (commented):
thanks for having such fantastic pr. your work is great and so complete. before diving into detailed review, i have some first thought to share:
- introducing expo-file-system as a dependency to expo-sqlite is not ideal for our version management. please see the package.json inline comment.
- we are planning to have `expo-file-system/next` in the upcoming expo sdk release. how do you think if we add the `getSharedContainerUriAsync` to the `expo-file-system/next` and only available from the next version? cc @aleqsio

IgorKhramtsov committed:
[sqlite] remove expo-file-system dependency

IgorKhramtsov committed:
Merge branch 'main' into ik/sqlite-custom-directory

IgorKhramtsov commented:
> we are planning to have expo-file-system/next in the upcoming expo sdk release. how do you think if we add the getSharedContainerUriAsync to the expo-file-system/next and only available from the next version? 

I'm working on it

IgorKhramtsov committed:
[file-system] add getSharedContainer func to file-system/next

IgorKhramtsov committed:
[file-system] remove getSharedContainerUriAsync from file-system

IgorKhramtsov committed:
[docs] update sqlite shared container guide

IgorKhramtsov committed:
chore: swiftlint

IgorKhramtsov committed:
[file-system] rename getSharedContainerUri to getSharedContainer

IgorKhramtsov commented:
@Kudo I've removed `getSharedContainerUriAsync` from `expo-file-system` and added `getSharedContainer` as a static function under `Directory` in the `expo-file-system/next`.

Kudo commented on packages/expo-modules-core/ios/FileSystemUtilities/FileSystemUtilities.swift:
```diff
@@ -56,6 +56,16 @@ public struct FileSystemUtilities {
       return permissionForInternalDirs
     }
+    let standardizedPath = path.standardizedFileURL.path
+    if let range = standardizedPath.range(of: "Containers/Shared/AppGroup") {
```
i'm sorry for late to review the pr. i have little concern of having this ios internal path here. also that's not ideal to maintain appGroupId from both entitlment and javascript calls. i have some idea to try out. are you open to let me have a pr to work on App Group beforehand (will list you as co-author for sure).

Kudo commented on packages/expo-sqlite/src/SQLiteDatabase.ts:
```diff
@@ -407,20 +407,63 @@ export class SQLiteDatabase {
+    return path.replace(/^\/+/, '');
+  }
+
+  return `${removeTrailingSlash(directory)}/${removeLeadingSlash(databaseName)}`;
```
let me double check if that works on windows

Kudo commented on packages/expo-sqlite/src/SQLiteDatabase.ts:
```diff
@@ -407,20 +407,63 @@ export class SQLiteDatabase {
 ): Promise<SQLiteDatabase> {
   const openOptions = options ?? {};
-  const nativeDatabase = new ExpoSQLite.NativeDatabase(databaseName, openOptions);
+  const databasePath = createDatabasePath(databaseName, resolveDbDirectory(directory));
```
there's too many occurrences of  `createDatabasePath(databaseName, resolveDbDirectory(directory));`. maybe we can move this to`pathUtils.ts` and have `function createDatabasePath(databaseName: string, directory?: string` there. then we can also do some js unit test there

Kudo reviewed (commented):
mostly looks great for expo-sqlite code changes. i will test windows support further and propose a pr for AppGroup change for expo-file-system

aleqsio commented on apps/test-suite/tests/FileSystemNext.ts:
```diff
@@ -283,5 +283,63 @@ export async function test({ describe, expect, it, ...t }) {
+    if (Platform.OS === 'ios') {
+      describe('App Group', () => {
+        const sharedContainerTestDir =
+          Directory.getSharedContainer('group.dev.expo.Payments').path + 'test/';
```
I'm working on an api for exposing common paths like `cache` and `documents` here:
https://github.com/expo/expo/pull/31467/

maybe this should be a property on `Paths`, like `Paths.getSharedContainer()`

IgorKhramtsov replied:
_all file-system changes are reverted_

aleqsio commented on packages/expo-file-system/ios/FileSystemExceptions.swift:
```diff
@@ -97,3 +97,9 @@ final class FailedToFindAssetException: GenericException<String> {
+
+final class CannotGetSharedContainer: GenericException<String> {
+  override var reason: String {
+    "Unable to open shared container with provided app group \(param). Check your entitlements"
```
`Check your entitlements`– could we make this more descriptive, like letting folks know which entitlements are missing or an FYI link?

IgorKhramtsov replied:
_all file-system changes are reverted_

aleqsio reviewed (approved):
Sorry for the late review, I like the FS part, but we'll need to update it to latest main and align with some recent changes. Happy to help out with this :)

Kudo commented:
> i have some idea to try out. are you open to let me have a pr to work on App Group beforehand (will list you as co-author for sure).

having the pr at #31525, basically the appGroupId is passing from native (as single source of truth), so the callsite doesn't need to specify appGroupId.

```ts
  /**
   * Gets the shared container directories for [Apple App Groups](https://developer.apple.com/documentation/xcode/configuring-app-groups).
   * @return Map of app group identifiers to shared directories.
   * @platform ios
   * @platform tvos
   */
  static getAppleSharedContainersAsync(): Promise<Record<string, Directory>>;
```

IgorKhramtsov commented:
> having the pr at https://github.com/expo/expo/pull/31525, basically the appGroupId is passing from native (as single source of truth), so the callsite doesn't need to specify appGroupId.

I checked your PRs, those changes look great! I will wait until they get merged to update my PR. As there are a few approvals on them already, I assume that will be quick.

Kudo commented:
> I checked your PRs, those changes look great! I will wait until they get merged to update my PR. As there are a few approvals on them already, I assume that will be quick.

thanks for waiting. all prs are landed on main. the new file-system api is `Paths.appleSharedContainers`. please help to rebase and i'd love to review this pr.

IgorKhramtsov committed:
revert: files-system changes

IgorKhramtsov committed:
Merge branch 'main' into ik/sqlite-custom-directory

IgorKhramtsov committed:
[expo-sqlite] fix test-suite

IgorKhramtsov committed:
[expo-sqlite] update docs

Kudo commented on apps/test-suite/tests/SQLite.ts:
```diff
@@ -1,7 +1,9 @@
 import { Asset } from 'expo-asset';
 import * as FS from 'expo-file-system';
+import { Directory } from 'expo-file-system/next';
```
```suggestion
import { Paths } from 'expo-file-system/next';
```

Kudo commented on apps/test-suite/tests/SQLite.ts:
```diff
@@ -882,6 +884,88 @@ CREATE TABLE IF NOT EXISTS users (user_id INTEGER PRIMARY KEY NOT NULL, name VAR
+    if (Platform.OS === 'ios') {
+      describe('iOS App Group', () => {
+        beforeAll(async () => {
+          const sharedContainerRoot = Directory.getSharedContainer('group.dev.expo.Payments');
+          const sharedContainerDir = sharedContainerRoot.path + 'SQLite';
```
```suggestion
          const sharedContainerRoot = Object.values(Paths.appleSharedContainers)?.[0];
          const sharedContainerDir = sharedContainerRoot.uri + 'SQLite';
```

Kudo commented on apps/test-suite/tests/SQLite.ts:
```diff
@@ -882,6 +884,88 @@ CREATE TABLE IF NOT EXISTS users (user_id INTEGER PRIMARY KEY NOT NULL, name VAR
+        });
+
+        it('should create and delete a database in a shared container', async () => {
+          const sharedContainerRoot = Directory.getSharedContainer('group.dev.expo.Payments');
+          const dbDirectory = sharedContainerRoot.path + 'SQLite';
```
```suggestion
          const sharedContainerRoot = Object.values(Paths.appleSharedContainers)?.[0];
          const dbDirectory = sharedContainerRoot.uri + 'SQLite';
```

Kudo commented on apps/test-suite/tests/SQLite.ts:
```diff
@@ -882,6 +884,88 @@ CREATE TABLE IF NOT EXISTS users (user_id INTEGER PRIMARY KEY NOT NULL, name VAR
+        });
+
+        it('should support internal importDatabaseFromAssetAsync without using expo-file-system', async () => {
+          const sharedContainerRoot = Directory.getSharedContainer('group.dev.expo.Payments');
+          const dbDirectory = sharedContainerRoot.path + 'SQLite';
```
```suggestion
          const sharedContainerRoot = Object.values(Paths.appleSharedContainers)?.[0];
          const dbDirectory = sharedContainerRoot.uri + 'SQLite';
```

Kudo commented on docs/pages/versions/unversioned/sdk/sqlite.mdx:
````diff
@@ -358,6 +359,54 @@ export default function App() {
+Use [`getSharedContainer`](filesystem-next/#getsharedcontainerappgroup) from the [`expo-file-system`](filesystem-next/) module to retrieve the path to the shared container:
+```tsx Using Shared Container for SQLite Database on iOS
+import { SQLiteProvider, defaultDatabaseDirectory } from 'expo-sqlite';
+import { Directory } from 'expo-file-system/next';
````
```suggestion
import { Paths } from 'expo-file-system/next';
```

Kudo commented on docs/pages/versions/unversioned/sdk/sqlite.mdx:
````diff
@@ -358,6 +359,54 @@ export default function App() {
+```
+</Step>
+<Step label="2">
+Use [`getSharedContainer`](filesystem-next/#getsharedcontainerappgroup) from the [`expo-file-system`](filesystem-next/) module to retrieve the path to the shared container:
````
```suggestion
Use [`appleSharedContainers`](filesystem-next/#applesharedcontainers) from the [`expo-file-system`](filesystem-next/) module to retrieve the path to the shared container:
```

Kudo commented on docs/pages/versions/unversioned/sdk/sqlite.mdx:
```diff
@@ -358,6 +359,54 @@ export default function App() {
+export default function App() {
+  const dbDirectory = useMemo(() => {
+    if (Platform.OS === 'ios') {
+      return Directory.getSharedContainer('group.com.myapp')?.path;
```
```suggestion
      return Object.values(Paths.appleSharedContainers)?.[0]?.uri;
```

Kudo reviewed (commented):
revise for `Paths.appleSharedContainers` api

IgorKhramtsov committed:
[expo-sqlite] move util functions to pathUtils.ts, add tests

IgorKhramtsov committed:
[expo-sqlite] update docs

IgorKhramtsov committed:
chore: kt formatting

IgorKhramtsov committed:
[expo-sqlite] add mocks for ensureDatabasePathExists funcs

IgorKhramtsov committed:
chore: docs formatting

amandeepmittal commented on docs/pages/versions/unversioned/sdk/sqlite.mdx:
```diff
@@ -358,6 +359,58 @@ export default function App() {
 
+### Sharing a database between apps/extensions (iOS)
+
+To share a database with other apps/extensions in the same App Group you can use shared containers. Follow these steps:
```
```suggestion
To share a database with other apps/extensions in the same App Group, you can use shared containers by following the steps below:
```

Update this to avoid using a fragmented second sentence so the readability is improved.

amandeepmittal commented on docs/pages/versions/unversioned/sdk/sqlite.mdx:
````diff
@@ -358,6 +359,58 @@ export default function App() {
+
+To share a database with other apps/extensions in the same App Group you can use shared containers. Follow these steps:
+
+<Step label="1">
+Configure the App Group in `app.json`:
+```json app.json
````
````suggestion
<Step label="1">

Configure the App Group in app config:

```json app.json
````

Nit: Using line breaks like the above here helps us to avoid any unknown markup rendering issues.
Let's change `app.json` to app config since this file can have different file formats in a project (example: app.config.js).

amandeepmittal commented on docs/pages/versions/unversioned/sdk/sqlite.mdx:
````diff
@@ -358,6 +359,58 @@ export default function App() {
+  }
+}
+```
+</Step>
+<Step label="2">
+Use [`Paths.appleSharedContainers`](filesystem-next/#applesharedcontainers) from the [`expo-file-system`](filesystem-next/) module to retrieve the path to the shared container:
````
```suggestion

</Step>

<Step label="2">

Use [`Paths.appleSharedContainers`](filesystem-next/#applesharedcontainers) from the [`expo-file-system`](filesystem-next/) library to retrieve the path to the shared container:
```

We prefer using ["library" over module](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#glossary) as per our writing style guide

IgorKhramtsov committed:
Merge branch 'main' into ik/sqlite-custom-directory

IgorKhramtsov committed:
update generated files after merge

IgorKhramtsov committed:
Update docs/pages/versions/unversioned/sdk/sqlite.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

IgorKhramtsov committed:
Update docs/pages/versions/unversioned/sdk/sqlite.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

IgorKhramtsov committed:
Update docs/pages/versions/unversioned/sdk/sqlite.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

IgorKhramtsov committed:
update mock to fix tests

amandeepmittal reviewed (approved):
Thanks for incorporating docs changes 👍 

Please resolve the lint error in `sqlite.mdx` file.

Kudo commented on packages/expo-sqlite/CHANGELOG.md:
```diff
@@ -9,6 +9,8 @@
 ### 🎉 New features
 
 - Added SQLCipher support. ([#30824](https://github.com/expo/expo/pull/30824), [#30825](https://github.com/expo/expo/pull/30825) by [@kudo](https://github.com/kudo))
+- Added key-value storage and compatible API with `@react-native-async-storage/async-storage`. ([#31596](https://github.com/expo/expo/pull/31596) by [@kudo](https://github.com/kudo))
```
```suggestion
```

Kudo reviewed (approved):
the changes looks all good to me. sorry again for the late review. please help to resolve conflicts and fix the lint error in sqlite.mdx. then we are good to merge. thanks for having the great change

IgorKhramtsov committed:
Merge remote-tracking branch 'origin' into ik/sqlite-custom-directory

IgorKhramtsov committed:
remove native next files after merge

IgorKhramtsov committed:
Update packages/expo-sqlite/CHANGELOG.md

Co-authored-by: Kudo Chien <ckchien@gmail.com>

IgorKhramtsov committed:
docs: fix lint error

IgorKhramtsov committed:
fix pathUtils.ios.ts after merge

IgorKhramtsov commented:
@Kudo done

Kudo reviewed (approved):
thanks again for the fantastic work 👏
